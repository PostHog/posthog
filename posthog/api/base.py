from rest_framework.pagination import CursorPagination as RestCursorPagination


class CursorPagination(RestCursorPagination):
    ordering = "-created_at"
    page_size = 100
