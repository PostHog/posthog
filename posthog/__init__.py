import sys
from unittest.mock import MagicMock
import deltalake.writer

# DLT imports the missing function `try_get_deltatable` from `deltalake` which
# it doesn't actually use but also no longer exists. So we can safely mock
# it to allow the app to startup without module errors
deltalake.writer.try_get_deltatable = MagicMock()  # type: ignore
sys.modules["deltalake.writer"] = deltalake.writer


# This will make sure the app is always imported when
# Django starts so that shared_task will use this app.
from posthog.celery import app as celery_app  # noqa: E402

__all__ = ("celery_app",)

# snowflake-connector-python tries to access a root folder which errors out in pods.
# This sets the snowflake home directory to a relative folder
import os  # noqa: E402

os.environ["SNOWFLAKE_HOME"] = "./.snowflake"
