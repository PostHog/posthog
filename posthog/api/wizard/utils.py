from __future__ import annotations

from google.genai.types import Type


def get_type_enum(json_type: str) -> Type:
    """Convert JSON Schema type to Gemini Type enum"""
    type_mapping = {
        "string": Type.STRING,
        "number": Type.NUMBER,
        "integer": Type.INTEGER,
        "boolean": Type.BOOLEAN,
        "array": Type.ARRAY,
        "object": Type.OBJECT,
    }
    return type_mapping.get(json_type, Type.STRING)


def get_field_mappings() -> dict[str, str]:
    """Get mappings from JSON Schema field names to Gemini schema field names"""
    return {
        "description": "description",
        "default": "default",
        "enum": "enum",
        "pattern": "pattern",
        "minLength": "min_length",
        "maxLength": "max_length",
        "minimum": "minimum",
        "maximum": "maximum",
        "minItems": "min_items",
        "maxItems": "max_items",
        "minProperties": "min_properties",
        "maxProperties": "max_properties",
        "required": "required",
        "nullable": "nullable",
        "title": "title",
        "example": "example",
    }


def convert_schema(schema: dict) -> dict:
    """Convert a single schema object from JSON Schema to Gemini format"""
    result = {}

    if "type" in schema:
        result["type"] = get_type_enum(schema["type"])

    field_mappings = get_field_mappings()
    for json_field, gemini_field in field_mappings.items():
        if json_field in schema:
            result[gemini_field] = schema[json_field]

    if "format" in schema:
        result["format"] = schema["format"]

    if "properties" in schema:
        result["properties"] = {key: convert_schema(value) for key, value in schema["properties"].items()}

    if "items" in schema:
        result["items"] = convert_schema(schema["items"])

    if "anyOf" in schema:
        result["any_of"] = [convert_schema(s) for s in schema["anyOf"]]

    return result


def json_schema_to_gemini_schema(json_schema: dict) -> dict:
    """Convert JSON Schema to Gemini Schema format"""
    actual_schema = json_schema.get("schema", json_schema)

    return convert_schema(actual_schema)
