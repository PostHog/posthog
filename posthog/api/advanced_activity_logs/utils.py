from datetime import datetime, timedelta
from typing import Optional

from django.utils import timezone

from posthog.constants import AvailableFeature
from posthog.models import Organization

from .constants import ADVANCED_ACTIVITY_LOGS_LOOKBACK_FALLBACK_LIMIT, ADVANCED_ACTIVITY_LOGS_LOOKBACK_FALLBACK_UNIT


def get_activity_log_lookback_restriction(organization: Organization) -> Optional[datetime]:
    """Get the lookback restriction date based on the AUDIT_LOGS feature."""
    audit_log_feature = organization.get_available_feature(AvailableFeature.AUDIT_LOGS)

    if not audit_log_feature:
        return None

    limit = audit_log_feature.get("limit")
    unit = audit_log_feature.get("unit")

    if limit is None or unit is None:
        limit = ADVANCED_ACTIVITY_LOGS_LOOKBACK_FALLBACK_LIMIT
        unit = ADVANCED_ACTIVITY_LOGS_LOOKBACK_FALLBACK_UNIT

    unit_lower = unit.lower()
    if unit_lower in ("day", "days"):
        delta = timedelta(days=limit)
    elif unit_lower in ("month", "months"):
        delta = timedelta(days=limit * 30)
    elif unit_lower in ("year", "years"):
        delta = timedelta(days=limit * 365)
    else:
        raise ValueError(f"Invalid unit: {unit}")

    return timezone.now() - delta
