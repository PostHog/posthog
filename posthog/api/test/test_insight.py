import json
from datetime import timedelta
from unittest.case import skip
from unittest.mock import patch

from django.utils import timezone
from freezegun import freeze_time
from rest_framework import status

from posthog.models import (
    Cohort,
    Dashboard,
    Event,
    Filter,
    Insight,
    Person,
    Team,
    User,
)
from posthog.models.organization import OrganizationMembership
from posthog.tasks.update_cache import update_dashboard_item_cache
from posthog.test.base import APIBaseTest, QueryMatchingTest, snapshot_postgres_queries


def insight_test_factory(event_factory, person_factory):
    class TestInsight(APIBaseTest, QueryMatchingTest):
        maxDiff = None

        CLASS_DATA_LEVEL_SETUP = False

        def test_get_insight_items(self):
            filter_dict = {
                "events": [{"id": "$pageview"}],
                "properties": [{"key": "$browser", "value": "Mac OS X"}],
            }

            Insight.objects.create(filters=Filter(data=filter_dict).to_dict(), team=self.team, created_by=self.user)

            # create without user
            Insight.objects.create(filters=Filter(data=filter_dict).to_dict(), team=self.team)

            response = self.client.get(f"/api/projects/{self.team.id}/insights/", data={"user": "true"}).json()

            self.assertEqual(len(response["results"]), 1)

        def test_get_saved_insight_items(self):
            filter_dict = {
                "events": [{"id": "$pageview"}],
                "properties": [{"key": "$browser", "value": "Mac OS X"}],
            }

            Insight.objects.create(
                filters=Filter(data=filter_dict).to_dict(), saved=True, team=self.team, created_by=self.user,
            )

            # create without saved
            Insight.objects.create(
                filters=Filter(data=filter_dict).to_dict(), team=self.team, created_by=self.user,
            )

            # create without user
            Insight.objects.create(filters=Filter(data=filter_dict).to_dict(), team=self.team)

            response = self.client.get(
                f"/api/projects/{self.team.id}/insights/", data={"saved": "true", "user": "true"}
            )
            self.assertEqual(response.status_code, status.HTTP_200_OK)

            self.assertEqual(len(response.json()["results"]), 1)
            self.assertEqual(len(response.json()["results"][0]["short_id"]), 8)

        def test_get_favorited_insight_items(self):
            filter_dict = {
                "events": [{"id": "$pageview"}],
                "properties": [{"key": "$browser", "value": "Mac OS X"}],
            }

            Insight.objects.create(
                filters=Filter(data=filter_dict).to_dict(), favorited=True, team=self.team, created_by=self.user,
            )

            # create without favorited
            Insight.objects.create(
                filters=Filter(data=filter_dict).to_dict(), team=self.team, created_by=self.user,
            )

            # create without user
            Insight.objects.create(filters=Filter(data=filter_dict).to_dict(), team=self.team)

            response = self.client.get(f"/api/projects/{self.team.id}/insights/?favorited=true&user=true")
            self.assertEqual(response.status_code, status.HTTP_200_OK)

            self.assertEqual(len(response.json()["results"]), 1)
            self.assertEqual((response.json()["results"][0]["favorited"]), True)

        def test_get_insight_by_short_id(self):
            filter_dict = {
                "events": [{"id": "$pageview"}],
            }

            Insight.objects.create(
                filters=Filter(data=filter_dict).to_dict(), team=self.team, short_id="12345678",
            )

            # Red herring: Should be ignored because it's not on the current team (even though the user has access)
            new_team = Team.objects.create(organization=self.organization)
            Insight.objects.create(
                filters=Filter(data=filter_dict).to_dict(), team=new_team, short_id="12345678",
            )

            response = self.client.get(f"/api/projects/{self.team.id}/insights/?short_id=12345678")
            self.assertEqual(response.status_code, status.HTTP_200_OK)

            self.assertEqual(len(response.json()["results"]), 1)
            self.assertEqual(response.json()["results"][0]["short_id"], "12345678")
            self.assertEqual(response.json()["results"][0]["filters"]["events"][0]["id"], "$pageview")

        def test_basic_results(self):
            """
            The `skip_results` query parameter can be passed so that only a list of objects is returned, without
            the actual query data. This can speed things up if it's not needed.
            """
            filter_dict = {
                "events": [{"id": "$pageview"}],
            }

            Insight.objects.create(
                filters=Filter(data=filter_dict).to_dict(), team=self.team, short_id="12345678",
            )
            Insight.objects.create(
                filters=Filter(data=filter_dict).to_dict(), team=self.team, saved=True,
            )

            response = self.client.get(f"/api/projects/{self.team.id}/insights/?basic=true")
            self.assertEqual(response.status_code, status.HTTP_200_OK)

            self.assertEqual(len(response.json()["results"]), 2)
            self.assertEqual(
                list(response.json()["results"][0].keys()),
                [
                    "id",
                    "short_id",
                    "name",
                    "filters",
                    "dashboard",
                    "color",
                    "description",
                    "last_refresh",
                    "refreshing",
                    "saved",
                    "updated_at",
                ],
            )

        @snapshot_postgres_queries
        def test_insights_does_not_nplus1(self):
            for i in range(20):
                user = User.objects.create(email=f"testuser{i}@posthog.com")
                OrganizationMembership.objects.create(user=user, organization=self.organization)
                dashboard = Dashboard.objects.create(name=f"Dashboard {i}", team=self.team)
                Insight.objects.create(
                    filters=Filter(data={"events": [{"id": "$pageview"}]}).to_dict(),
                    team=self.team,
                    short_id=f"insight{i}",
                    dashboard=dashboard,
                    created_by=user,
                )

            # 4 for request overhead (django sessions/auth), then item count + items + dashboards + users
            with self.assertNumQueries(8):
                response = self.client.get(f"/api/projects/{self.team.id}/insights")
            self.assertEqual(response.status_code, status.HTTP_200_OK)
            self.assertEqual(len(response.json()["results"]), 20)

        def test_create_insight_items(self):
            # Make sure the endpoint works with and without the trailing slash
            response = self.client.post(
                f"/api/projects/{self.team.id}/insights",
                data={
                    "filters": {
                        "events": [{"id": "$pageview"}],
                        "properties": [{"key": "$browser", "value": "Mac OS X"}],
                        "date_from": "-90d",
                    },
                },
            )
            self.assertEqual(response.status_code, status.HTTP_201_CREATED)
            self.assertEqual(response.json()["description"], None)
            self.assertEqual(response.json()["tags"], [])

            objects = Insight.objects.all()
            self.assertEqual(len(objects), 1)
            self.assertEqual(objects[0].filters["events"][0]["id"], "$pageview")
            self.assertEqual(objects[0].filters["date_from"], "-90d")
            self.assertEqual(len(objects[0].short_id), 8)

        def test_update_insight(self):
            insight = Insight.objects.create(team=self.team, name="special insight", created_by=self.user,)
            response = self.client.patch(
                f"/api/projects/{self.team.id}/insights/{insight.id}",
                {
                    "name": "insight new name",
                    "tags": ["official", "engineering"],
                    "description": "Internal system metrics.",
                },
            )
            self.assertEqual(response.status_code, status.HTTP_200_OK)

            response_data = response.json()
            self.assertEqual(response_data["name"], "insight new name")
            self.assertEqual(response_data["created_by"]["distinct_id"], self.user.distinct_id)
            self.assertEqual(response_data["description"], "Internal system metrics.")
            self.assertEqual(response_data["tags"], ["official", "engineering"])

            insight.refresh_from_db()
            self.assertEqual(insight.name, "insight new name")
            self.assertEqual(insight.tags, ["official", "engineering"])

        @skip("Compatibility issue caused by test account filters")
        def test_update_insight_filters(self):
            insight = Insight.objects.create(
                team=self.team,
                name="insight with custom filters",
                created_by=self.user,
                filters={"events": [{"id": "$pageview"}]},
            )

            for custom_name, expected_name in zip(
                ["Custom filter", 100, "", "  ", None], ["Custom filter", "100", None, None, None]
            ):
                response = self.client.patch(
                    f"/api/projects/{self.team.id}/insights/{insight.id}",
                    {"filters": {"events": [{"id": "$pageview", "custom_name": custom_name}]}},
                )

                self.assertEqual(response.status_code, status.HTTP_200_OK)

                response_data = response.json()
                self.assertEqual(response_data["filters"]["events"][0]["custom_name"], expected_name)
                insight.refresh_from_db()
                self.assertEqual(insight.filters["events"][0]["custom_name"], expected_name)

        def test_save_new_funnel(self):

            dashboard = Dashboard.objects.create(name="My Dashboard", team=self.team)

            response = self.client.post(
                f"/api/projects/{self.team.id}/insights",
                data={
                    "filters": {
                        "insight": "FUNNELS",
                        "events": [
                            {
                                "id": "$pageview",
                                "math": None,
                                "name": "$pageview",
                                "type": "events",
                                "order": 0,
                                "properties": [],
                                "math_property": None,
                            },
                            {
                                "id": "$rageclick",
                                "math": None,
                                "name": "$rageclick",
                                "type": "events",
                                "order": 2,
                                "properties": [],
                                "math_property": None,
                            },
                        ],
                        "display": "FunnelViz",
                        "interval": "day",
                        "date_from": "-30d",
                        "actions": [],
                        "new_entity": [],
                        "layout": "horizontal",
                    },
                    "name": "My Funnel One",
                    "dashboard": dashboard.pk,
                },
            )
            self.assertEqual(response.status_code, status.HTTP_201_CREATED)

            objects = Insight.objects.all()
            self.assertEqual(len(objects), 1)
            self.assertEqual(objects[0].filters["events"][1]["id"], "$rageclick")
            self.assertEqual(objects[0].filters["display"], "FunnelViz")
            self.assertEqual(objects[0].filters["interval"], "day")
            self.assertEqual(objects[0].filters["date_from"], "-30d")
            self.assertEqual(objects[0].filters["layout"], "horizontal")
            self.assertEqual(len(objects[0].short_id), 8)

        @patch("posthog.api.insight.update_dashboard_item_cache", wraps=update_dashboard_item_cache)
        def test_insight_refreshing(self, spy_update_dashboard_item_cache):
            with freeze_time("2012-01-14T03:21:34.000Z"):
                event_factory(team=self.team, event="$pageview", distinct_id="1")
                event_factory(team=self.team, event="$pageview", distinct_id="2")

            with freeze_time("2012-01-15T04:01:34.000Z"):
                response = self.client.post(
                    f"/api/projects/{self.team.id}/insights", data={"filters": {"events": [{"id": "$pageview"}]}}
                ).json()
                self.assertEqual(response["last_refresh"], None)

                response = self.client.get(
                    f"/api/projects/{self.team.id}/insights/{response['id']}/?refresh=true"
                ).json()
                self.assertEqual(spy_update_dashboard_item_cache.call_count, 1)
                self.assertEqual(response["result"][0]["data"], [0, 0, 0, 0, 0, 0, 2, 0])
                self.assertEqual(response["last_refresh"], "2012-01-15T04:01:34Z")
                self.assertEqual(response["updated_at"], "2012-01-15T04:01:34Z")

            with freeze_time("2012-01-15T05:01:34.000Z"):
                event_factory(team=self.team, event="$pageview", distinct_id="1")
                response = self.client.get(
                    f"/api/projects/{self.team.id}/insights/{response['id']}/?refresh=true"
                ).json()
                self.assertEqual(spy_update_dashboard_item_cache.call_count, 2)
                self.assertEqual(response["result"][0]["data"], [0, 0, 0, 0, 0, 0, 2, 1])
                self.assertEqual(response["last_refresh"], "2012-01-15T05:01:34Z")
                self.assertEqual(response["updated_at"], "2012-01-15T04:01:34Z")  # did not change

            with freeze_time("2012-01-25T05:01:34.000Z"):
                response = self.client.get(f"/api/projects/{self.team.id}/insights/{response['id']}/").json()
                self.assertEqual(spy_update_dashboard_item_cache.call_count, 2)
                self.assertEqual(response["last_refresh"], None)
                self.assertEqual(response["updated_at"], "2012-01-15T04:01:34Z")  # did not change

        # BASIC TESTING OF ENDPOINTS. /queries as in depth testing for each insight

        def test_insight_trends_basic(self):
            with freeze_time("2012-01-14T03:21:34.000Z"):
                event_factory(team=self.team, event="$pageview", distinct_id="1")
                event_factory(team=self.team, event="$pageview", distinct_id="2")

            with freeze_time("2012-01-15T04:01:34.000Z"):
                response = self.client.get(
                    f"/api/projects/{self.team.id}/insights/trend/?events={json.dumps([{'id': '$pageview'}])}"
                ).json()

            self.assertEqual(response["result"][0]["count"], 2)
            self.assertEqual(response["result"][0]["action"]["name"], "$pageview")

        def test_nonexistent_cohort_is_handled(self):
            response_nonexistent_property = self.client.get(
                f"/api/projects/{self.team.id}/insights/trend/?events={json.dumps([{'id': '$pageview'}])}&properties={json.dumps([{'type':'event','key':'foo','value':'barabarab'}])}"
            )
            response_nonexistent_cohort = self.client.get(
                f"/api/projects/{self.team.id}/insights/trend/?events={json.dumps([{'id': '$pageview'}])}&properties={json.dumps([{'type':'cohort','key':'id','value':2137}])}"
            )  # This should not throw an error, just act like there's no event matches

            response_nonexistent_property_data = response_nonexistent_property.json()
            response_nonexistent_cohort_data = response_nonexistent_cohort.json()
            response_nonexistent_property_data.pop("last_refresh")
            response_nonexistent_cohort_data.pop("last_refresh")
            self.assertEntityResponseEqual(
                response_nonexistent_property_data["result"], response_nonexistent_cohort_data["result"]
            )  # Both cases just empty

        def test_cohort_without_match_group_works(self):
            whatever_cohort_without_match_groups = Cohort.objects.create(team=self.team)

            response_nonexistent_property = self.client.get(
                f"/api/projects/{self.team.id}/insights/trend/?events={json.dumps([{'id': '$pageview'}])}&properties={json.dumps([{'type':'event','key':'foo','value':'barabarab'}])}"
            )
            response_cohort_without_match_groups = self.client.get(
                f"/api/projects/{self.team.id}/insights/trend/?events={json.dumps([{'id':'$pageview'}])}&properties={json.dumps([{'type':'cohort','key':'id','value':whatever_cohort_without_match_groups.pk}])}"
            )  # This should not throw an error, just act like there's no event matches

            self.assertEqual(response_nonexistent_property.status_code, 200)
            response_nonexistent_property_data = response_nonexistent_property.json()
            response_cohort_without_match_groups_data = response_cohort_without_match_groups.json()
            response_nonexistent_property_data.pop("last_refresh")
            response_cohort_without_match_groups_data.pop("last_refresh")
            self.assertEntityResponseEqual(
                response_nonexistent_property_data["result"], response_cohort_without_match_groups_data["result"]
            )  # Both cases just empty

        def test_precalculated_cohort_works(self):
            person_factory(team=self.team, distinct_ids=["person_1"], properties={"foo": "bar"})

            whatever_cohort: Cohort = Cohort.objects.create(
                id=113,
                team=self.team,
                groups=[{"properties": [{"type": "person", "key": "foo", "value": "bar", "operator": "exact"}]}],
                last_calculation=timezone.now(),
            )
            whatever_cohort.calculate_people()
            whatever_cohort.calculate_people_ch()

            with self.settings(USE_PRECALCULATED_CH_COHORT_PEOPLE=True):  # Normally this is False in tests
                response_user_property = self.client.get(
                    f"/api/projects/{self.team.id}/insights/trend/?events={json.dumps([{'id': '$pageview'}])}&properties={json.dumps([{'type':'person','key':'foo','value':'bar'}])}"
                )
                response_precalculated_cohort = self.client.get(
                    f"/api/projects/{self.team.id}/insights/trend/?events={json.dumps([{'id':'$pageview'}])}&properties={json.dumps([{'type':'cohort','key':'id','value':113}])}"
                )

            self.assertEqual(response_precalculated_cohort.status_code, 200)
            response_user_property_data = response_user_property.json()
            response_precalculated_cohort_data = response_precalculated_cohort.json()
            response_user_property_data.pop("last_refresh")
            response_precalculated_cohort_data.pop("last_refresh")

            self.assertEntityResponseEqual(
                response_user_property_data["result"], response_precalculated_cohort_data["result"]
            )

        def test_insight_trends_compare(self):
            with freeze_time("2012-01-14T03:21:34.000Z"):
                for i in range(25):
                    event_factory(
                        team=self.team, event="$pageview", distinct_id="1", properties={"$some_property": f"value{i}"},
                    )

            with freeze_time("2012-01-15T04:01:34.000Z"):
                response = self.client.get(
                    f"/api/projects/{self.team.id}/insights/trend/",
                    data={"events": json.dumps([{"id": "$pageview"}]), "compare": "true",},
                )
            self.assertEqual(response.status_code, status.HTTP_200_OK)
            result = response.json()
            self.assertEqual(len(result["result"]), 2)
            self.assertEqual(result["result"][0]["compare_label"], "current")
            self.assertEqual(result["result"][1]["compare_label"], "previous")

        def test_insight_trends_breakdown_pagination(self):
            with freeze_time("2012-01-14T03:21:34.000Z"):
                for i in range(25):

                    event_factory(
                        team=self.team, event="$pageview", distinct_id="1", properties={"$some_property": f"value{i}"},
                    )

            with freeze_time("2012-01-15T04:01:34.000Z"):
                response = self.client.get(
                    f"/api/projects/{self.team.id}/insights/trend/",
                    data={
                        "events": json.dumps([{"id": "$pageview"}]),
                        "breakdown": "$some_property",
                        "breakdown_type": "event",
                    },
                )
            self.assertEqual(response.status_code, status.HTTP_200_OK)
            self.assertIn("offset=20", response.json()["next"])

        def test_insight_paths_basic(self):
            person_factory(team=self.team, distinct_ids=["person_1"])
            event_factory(
                properties={"$current_url": "/", "test": "val"},
                distinct_id="person_1",
                event="$pageview",
                team=self.team,
            )
            event_factory(
                properties={"$current_url": "/about", "test": "val"},
                distinct_id="person_1",
                event="$pageview",
                team=self.team,
            )

            person_factory(team=self.team, distinct_ids=["dontcount"])
            event_factory(
                properties={"$current_url": "/", "test": "val"},
                distinct_id="dontcount",
                event="$pageview",
                team=self.team,
            )
            event_factory(
                properties={"$current_url": "/about", "test": "val"},
                distinct_id="dontcount",
                event="$pageview",
                team=self.team,
            )

            get_response = self.client.get(
                f"/api/projects/{self.team.id}/insights/path",
                data={"properties": json.dumps([{"key": "test", "value": "val"}]),},
            ).json()
            post_response = self.client.post(
                f"/api/projects/{self.team.id}/insights/path", {"properties": [{"key": "test", "value": "val"}],}
            ).json()
            self.assertEqual(len(get_response["result"]), 1)
            self.assertEqual(len(post_response["result"]), 1)

        def test_insight_funnels_basic_post(self):
            person_factory(team=self.team, distinct_ids=["1"])
            event_factory(team=self.team, event="user signed up", distinct_id="1")
            event_factory(team=self.team, event="user did things", distinct_id="1")
            response = self.client.post(
                f"/api/projects/{self.team.id}/insights/funnel/",
                {
                    "events": [
                        {"id": "user signed up", "type": "events", "order": 0},
                        {"id": "user did things", "type": "events", "order": 1},
                    ],
                    "funnel_window_days": 14,
                },
            ).json()

            # clickhouse funnels don't have a loading system
            self.assertEqual(len(response["result"]), 2)
            self.assertEqual(response["result"][0]["name"], "user signed up")
            self.assertEqual(response["result"][0]["count"], 1)
            self.assertEqual(response["result"][1]["name"], "user did things")
            self.assertEqual(response["result"][1]["count"], 1)

        # Tests backwards-compatibility when we changed GET to POST | GET
        def test_insight_funnels_basic_get(self):
            event_factory(team=self.team, event="user signed up", distinct_id="1")
            event_factory(team=self.team, event="user did things", distinct_id="1")
            response = self.client.get(
                f"/api/projects/{self.team.id}/insights/funnel/?funnel_window_days=14&events={json.dumps([{'id': 'user signed up', 'type': 'events', 'order': 0},{'id': 'user did things', 'type': 'events', 'order': 1},])}"
            ).json()

            # clickhouse funnels don't have a loading system
            self.assertEqual(len(response["result"]), 2)
            self.assertEqual(response["result"][0]["name"], "user signed up")
            self.assertEqual(response["result"][1]["name"], "user did things")

        def test_insight_retention_basic(self):
            person_factory(team=self.team, distinct_ids=["person1"], properties={"email": "person1@test.com"})
            event_factory(
                team=self.team, event="$pageview", distinct_id="person1", timestamp=timezone.now() - timedelta(days=11),
            )

            event_factory(
                team=self.team, event="$pageview", distinct_id="person1", timestamp=timezone.now() - timedelta(days=10),
            )
            response = self.client.get(f"/api/projects/{self.team.id}/insights/retention/",).json()

            self.assertEqual(len(response["result"]), 11)

        def test_insight_with_specified_token(self):
            _, _, user2 = User.objects.bootstrap("Test", "team2@posthog.com", None)
            assert user2.team is not None
            assert self.team is not None
            assert self.user.team is not None

            self.assertNotEqual(user2.team.id, self.team.id)
            self.client.force_login(self.user)

            person_factory(team=self.team, distinct_ids=["person1"], properties={"email": "person1@test.com"})

            event_factory(
                team=self.team, event="$pageview", distinct_id="person1", timestamp=timezone.now() - timedelta(days=6),
            )

            event_factory(
                team=self.team, event="$pageview", distinct_id="person1", timestamp=timezone.now() - timedelta(days=5),
            )

            events_filter = json.dumps([{"id": "$pageview"}])

            response_team1 = self.client.get(f"/api/projects/{self.team.id}/insights/trend/?events={events_filter}")
            response_team1_token = self.client.get(
                f"/api/projects/{self.team.id}/insights/trend/?events={events_filter}&token={self.user.team.api_token}"
            )

            self.client.force_login(user2)
            response_team2 = self.client.get(
                f"/api/projects/{self.team.id}/insights/trend/?events={events_filter}",
                data={"token": user2.team.api_token},
            )

            self.assertEqual(response_team1.status_code, 200)
            self.assertEqual(response_team2.status_code, 200)
            self.assertEqual(response_team1.json()["result"], response_team1_token.json()["result"])
            self.assertNotEqual(len(response_team1.json()["result"]), len(response_team2.json()["result"]))

            response_invalid_token = self.client.get(f"/api/projects/{self.team.id}/insights/trend?token=invalid")
            self.assertEqual(response_invalid_token.status_code, 401)

    return TestInsight
