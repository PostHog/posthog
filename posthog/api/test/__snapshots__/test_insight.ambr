# name: TestInsight.test_insights_does_not_nplus1
  '
  SELECT "posthog_user"."id",
         "posthog_user"."password",
         "posthog_user"."last_login",
         "posthog_user"."first_name",
         "posthog_user"."last_name",
         "posthog_user"."is_staff",
         "posthog_user"."is_active",
         "posthog_user"."date_joined",
         "posthog_user"."uuid",
         "posthog_user"."current_organization_id",
         "posthog_user"."current_team_id",
         "posthog_user"."email",
         "posthog_user"."temporary_token",
         "posthog_user"."distinct_id",
         "posthog_user"."email_opt_in",
         "posthog_user"."anonymize_data",
         "posthog_user"."toolbar_mode",
         "posthog_user"."events_column_config"
  FROM "posthog_user"
  WHERE "posthog_user"."id" = 2
  LIMIT 21
  '
---
# name: TestInsight.test_insights_does_not_nplus1.1
  '
  SELECT "posthog_team"."id",
         "posthog_team"."uuid",
         "posthog_team"."organization_id",
         "posthog_team"."api_token",
         "posthog_team"."app_urls",
         "posthog_team"."name",
         "posthog_team"."slack_incoming_webhook",
         "posthog_team"."created_at",
         "posthog_team"."updated_at",
         "posthog_team"."anonymize_ips",
         "posthog_team"."completed_snippet_onboarding",
         "posthog_team"."ingested_event",
         "posthog_team"."session_recording_opt_in",
         "posthog_team"."signup_token",
         "posthog_team"."is_demo",
         "posthog_team"."access_control",
         "posthog_team"."test_account_filters",
         "posthog_team"."path_cleaning_filters",
         "posthog_team"."timezone",
         "posthog_team"."data_attributes",
         "posthog_team"."correlation_config",
         "posthog_team"."session_recording_retention_period_days",
         "posthog_team"."plugins_opt_in",
         "posthog_team"."opt_out_capture",
         "posthog_team"."event_names",
         "posthog_team"."event_names_with_usage",
         "posthog_team"."event_properties",
         "posthog_team"."event_properties_with_usage",
         "posthog_team"."event_properties_numerical"
  FROM "posthog_team"
  WHERE "posthog_team"."id" = 2
  LIMIT 21
  '
---
# name: TestInsight.test_insights_does_not_nplus1.2
  '
  SELECT "posthog_organizationmembership"."id",
         "posthog_organizationmembership"."organization_id",
         "posthog_organizationmembership"."user_id",
         "posthog_organizationmembership"."level",
         "posthog_organizationmembership"."joined_at",
         "posthog_organizationmembership"."updated_at",
         "posthog_organization"."id",
         "posthog_organization"."name",
         "posthog_organization"."slug",
         "posthog_organization"."created_at",
         "posthog_organization"."updated_at",
         "posthog_organization"."domain_whitelist",
         "posthog_organization"."setup_section_2_completed",
         "posthog_organization"."personalization",
         "posthog_organization"."plugins_access_level",
         "posthog_organization"."available_features",
         "posthog_organization"."for_internal_metrics",
         "posthog_organization"."is_member_join_email_enabled"
  FROM "posthog_organizationmembership"
  INNER JOIN "posthog_organization" ON ("posthog_organizationmembership"."organization_id" = "posthog_organization"."id")
  WHERE ("posthog_organizationmembership"."organization_id" = '00000000-0000-0000-0000-000000000000'::uuid
         AND "posthog_organizationmembership"."user_id" = 2)
  LIMIT 21
  '
---
# name: TestInsight.test_insights_does_not_nplus1.3
  '
  SELECT COUNT(*) AS "__count"
  FROM "posthog_dashboarditem"
  WHERE ("posthog_dashboarditem"."team_id" = 2
         AND NOT "posthog_dashboarditem"."deleted")
  '
---
# name: TestInsight.test_insights_does_not_nplus1.4
  '
  SELECT "posthog_dashboarditem"."id",
         "posthog_dashboarditem"."dashboard_id",
         "posthog_dashboarditem"."dive_dashboard_id",
         "posthog_dashboarditem"."name",
         "posthog_dashboarditem"."description",
         "posthog_dashboarditem"."team_id",
         "posthog_dashboarditem"."filters",
         "posthog_dashboarditem"."filters_hash",
         "posthog_dashboarditem"."order",
         "posthog_dashboarditem"."deleted",
         "posthog_dashboarditem"."saved",
         "posthog_dashboarditem"."created_at",
         "posthog_dashboarditem"."layouts",
         "posthog_dashboarditem"."color",
         "posthog_dashboarditem"."last_refresh",
         "posthog_dashboarditem"."refreshing",
         "posthog_dashboarditem"."created_by_id",
         "posthog_dashboarditem"."is_sample",
         "posthog_dashboarditem"."short_id",
         "posthog_dashboarditem"."updated_at",
         "posthog_dashboarditem"."tags",
         "posthog_dashboarditem"."favorited",
         "posthog_dashboarditem"."refresh_attempt"
  FROM "posthog_dashboarditem"
  WHERE ("posthog_dashboarditem"."team_id" = 2
         AND NOT "posthog_dashboarditem"."deleted")
  ORDER BY "posthog_dashboarditem"."order" ASC
  LIMIT 100
  '
---
# name: TestInsight.test_insights_does_not_nplus1.5
  '
  SELECT "posthog_dashboard"."id",
         "posthog_dashboard"."name",
         "posthog_dashboard"."description",
         "posthog_dashboard"."team_id",
         "posthog_dashboard"."pinned",
         "posthog_dashboard"."created_at",
         "posthog_dashboard"."created_by_id",
         "posthog_dashboard"."deleted",
         "posthog_dashboard"."share_token",
         "posthog_dashboard"."is_shared",
         "posthog_dashboard"."last_accessed_at",
         "posthog_dashboard"."filters",
         "posthog_dashboard"."creation_mode",
         "posthog_dashboard"."tags"
  FROM "posthog_dashboard"
  WHERE "posthog_dashboard"."id" IN (1,
                                     2,
                                     3,
                                     4,
                                     5 /* ... */)
  '
---
# name: TestInsight.test_insights_does_not_nplus1.6
  '
  SELECT "posthog_user"."id",
         "posthog_user"."password",
         "posthog_user"."last_login",
         "posthog_user"."first_name",
         "posthog_user"."last_name",
         "posthog_user"."is_staff",
         "posthog_user"."is_active",
         "posthog_user"."date_joined",
         "posthog_user"."uuid",
         "posthog_user"."current_organization_id",
         "posthog_user"."current_team_id",
         "posthog_user"."email",
         "posthog_user"."temporary_token",
         "posthog_user"."distinct_id",
         "posthog_user"."email_opt_in",
         "posthog_user"."anonymize_data",
         "posthog_user"."toolbar_mode",
         "posthog_user"."events_column_config"
  FROM "posthog_user"
  WHERE "posthog_user"."id" IN (1,
                                2,
                                3,
                                4,
                                5 /* ... */)
  '
---
